rfcs/0136-stabilize-incrementally.md

```diff
@@ -0,0 +1,199 @@
+This is the opposite of how experimental features are supposed to leisurely incubate with both plenty of flexibility to change them, and little rush to stabilize them until they feel ready.
+But, there is little we can do about this at this point, an this RFC recognizes that fact and does *not* try to resist it.
+
+But, we have crossed that Rubicon and there is no turning back; this RFC *doesn't* attempt to change Flakes or the command line.
+
+# Detailed design
+[design]: #detailed-design
+
+Step 1 is technical work, with a self-imposed deadline so we can be sure it doesn't delay stabilization too long.
+The remaining steps are stabilization steps.
+For each of them, a separate RFC or other discussion medium will describe the new interfaces to be stabilized, and solicit feedback.
+
+## Step 0: Stabilize the store-only installable-free CLI
+
+There are certain commands like `nix store gc`, `nix store ping` that do not take any positional arguments.
+As @roberth elsewhere pointed out, because these commands have so few degrees of freedom, they are some of the easiest to stabilize --- there is simply less to pour over and possibly bikeshed.
+
+We can start stabilizing them right away, either in a batch, or one by one.
+
+## Step 1: Split out a store-only Nix, with deadline
+
+This is the same as step one from [RFC 134](https://github.com/NixOS/rfcs/pull/134), upon which this RFC depends.
+
+A deadline is imposed to get that complete, so agreement on this stabilization plan is guaranteed not to add more than a small bounded delay as opposed to stabilizing everything immediately.
+
+## Step 2: Stabilize the store-only Nix CLI
+
+If the deadline for step 1 is met, then we stabilize *just* the command-line interface of the store-only Nix command.
+
+This is a small portion of the total amount of interface we have to stabilize, and that will hopefully yield a narrow and more focused discussion.
+Yet it will still offer some interesting topics to discus, such as:
+
+- logging
+- store paths on `stdout` at end of build?
+- Should commands like `show-derivation` use `--json` by default?
+- Flat vs hierarchical commands
+- is `--derivation` a good flag?
+
+## Step 3: Stabilize the rest of the CLI, without Flakes
+
+If the deadline for step 1 is *not* met, we jump directly to this step, skipping step 2.
+
+This is the rest off the new CLI, without flakes.
+Unlike the store-only Nix command which has yet to be implemented, this is easy to visualize today by enabling the `nix-command` feature without the `flakes` feature.
+This is a chance to discuss topics like:
+
+- Is `--file` good, or should we be able to specify something like `file#attribute` to mix files and attributes? (Without using Flakes.)
+- Should all outputs be selected if one writes `foo.dev`?
+- How can `nix repl` have a more normal CLI?
+
+## Step 4: Stabilize Flakes itself
+
+Finally, with the other less controversial interfaces stabilized, we can tackle the Flakes itself, the one remainder.
```

Comment:
Is this thought of as a all-in-one thing (flake file format, registry, installables format, ..) or will we have seperate discussions on these topics? The old RFC covered a wide range of things (which wasn't great) so we might want do keep this in mind.

Reply:
Yeah I don't mean for this RFC to *preclude* further splitting up and incremental stabilizing, just obligate us to to *at least* try stabilizing these things first.